### PR TITLE
Base Image Backup/Restore

### PR DESCRIPTION
**Pull request by LLParse**

### CODE REVIEW

**Comment by yasker** [comment]
Can you try https://github.com/rancher/longhorn-tests/pull/94 if the `test_backup_with_base_image` can pass?

File: api/snapshot.go [hunk 1]
```diff
@@ -119,7 +119,17 @@ func (s *Server) SnapshotBackup(w http.ResponseWriter, req *http.Request) (err e
...
 
-	return s.m.BackupSnapshot(input.Name, nil, volName)
+	vol, err := s.m.Get(volName)
+	if err != nil {
+		return errors.Wrap(err, "unable to get volume")
+	}
+
+	labels := make(map[string]string)
+	if vol.Spec.BaseImage != "" {
+		labels["ranchervm-base-image"] = vol.Spec.BaseImage
```
Comment (LLParse) [inline api/snapshot.go]: label should be a constant

**Comment by LLParse** [comment]
Test passed.

File: controller/replica_controller.go [hunk 2]
```diff
@@ -294,6 +294,10 @@ func (rc *ReplicaController) getReadinessProbeFailureThreshold(r *longhorn.Repli
 	return int32(r.Spec.VolumeSize / replicaReadinessProbeMinimalRestoreRate / replicaReadinessProbePeriodSeconds)
 }
 
+func singleQuotes(static string) string {
```
Comment (yasker) [inline controller/replica_controller.go]: Got an issue?

File: api/snapshot.go (hunk 1 above)
Comment (yasker) [reply api/snapshot.go]: yeah

File: controller/replica_controller.go (hunk 2 above)
Comment (LLParse) [reply controller/replica_controller.go]: Yeah, the issue is the way we construct command line arguments with unescaped user input which happens to have hypens and confuses the CLI parser. Solution is to treat user input as shell unevaluated string.

**Pull request merged**